### PR TITLE
Added `Dictionary.mapKeys` and `Dictionary.compactMapKeys`

### PR DESCRIPTION
**Pull request by NachoSoto**
Fixes [CSDK-105].

This came up in https://github.com/RevenueCat/purchases-ios/pull/1534#discussion_r891433648 and will simplify some of the code in #1727.

[CSDK-105]: https://revenuecats.atlassian.net/browse/CSDK-105?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: Tests/UnitTests/FoundationExtensions/DictionaryExtensionsTests.swift [hunk 1]
```diff
@@ -133,4 +142,87 @@ class DictionaryExtensionsTests: TestCase {
...
+            4: 4,
+            5: 5
+        ]
+
+        let output = input.compactMapKeys {
+            $0.isMultiple(of: 2)
+                ? String($0)
+                : nil
+        }
+        expect(output) == [
```
Comment (NachoSoto) [inline Tests/UnitTests/FoundationExtensions/DictionaryExtensionsTests.swift:222]: Wait how is this test passing 🤨

File: Tests/UnitTests/FoundationExtensions/DictionaryExtensionsTests.swift [hunk 2]
```diff
@@ -133,4 +142,87 @@ class DictionaryExtensionsTests: TestCase {
...
+        expect([1: "test"].compactMapKeys(self.transformer)) == [
+            "1": "test"
+        ]
+    }
+
+    func testCompactMapDictionaryResultingInEmptyDictionary() {
+        expect([1: "test"].compactMapKeys { _ in nil }) == [:]
+    }
+
+    func tetCompactMapKeys() {
```
Comment (NachoSoto) [inline Tests/UnitTests/FoundationExtensions/DictionaryExtensionsTests.swift:208]: OMG 🤦‍♂️

Comment (aboedo) [reply Tests/UnitTests/FoundationExtensions/DictionaryExtensionsTests.swift:208]: 😂

**Comment by NachoSoto** [comment]
Fixed!

**Review by aboedo — APPROVED** [review]

**Pull request merged**